src/squlearn/encoding_circuit/circuit_library/qcnn_encoding_circuit.py

```diff
@@ -0,0 +1,708 @@
+from qiskit.circuit import ParameterVector
+from qiskit.converters import circuit_to_gate, circuit_to_instruction
+
+from squlearn.encoding_circuit.encoding_circuit_base import EncodingCircuitBase
+from squlearn.encoding_circuit.circuit_library.param_z_feature_map import ParamZFeatureMap
+from squlearn.observables import CustomObservable
+
+
+class QCNNEncodingCircuit(EncodingCircuitBase):
+
+    """
+    Encoding circuit for quantum convolutional neural networks (QCNN).
+
+    The structure is inspired by classical convolutional neural networks. The number of active
+    qubits reduces with each layer. The design idea was initially proposed in reference [1].
+
+    Args:
+        num_qubits (int): Number of initial qubits of the QCNN encoding circuit.
+        num_features (int): Dimension of the feature vector.
+            By default this is 0, so a feature map must be provided.
+            If the number of features is bigger then 0,
+            then in the get_circuit function a ZFeatureMap is built to encode the features.
+        default (bool): If True, the default circuit is built.
+
+    References
+    -----------
+    [1]: `Cong, I., Choi, S. & Lukin, M.D. Quantum convolutional neural networks. Nat. Phys. 15,
+    1273–1278 (2019). <https://doi.org/10.1038/s41567-019-0648-8>`_
+    """
+
+    def __init__(self, num_qubits: int = 0, num_features: int = 0, default: bool = False) -> None:
+        super().__init__(num_qubits, num_features)
+        self._num_parameters = 0
+        self._left_qubits = [i for i in range(num_qubits)]
+        self._operations_list = []
+        self._default = default
+        if default:
+            if num_qubits == 0:
+                print("To generate a default circuit provide a number of qubits > 0.")
+            else:
+                self.default_circuit()
+
+    @property
+    def num_parameters(self) -> int:
+        """Returns the number of trainable parameters of the current encoding circuit."""
+        return self._num_parameters
+
+    @property
+    def left_qubits(self) -> list:
+        """Returns the qubits which one can operate on in the current circuit."""
+        return self._left_qubits
+
+    @property
+    def operations_list(self) -> list:
+        """Returns the list of operators currently acting on the encoding circuit."""
+        return self._operations_list
+
+    def set_params(self, **params):
+        """
+        Sets value of the encoding circuit hyper-parameters.
+
+        If the number of qubits is reduced and the supplied pooling gates do not fit to this,
+        this will throw a error in the troubling layer.
+
+        Args:
+            params: Hyper-parameters (num_qubits or num_features)
+                and their values, e.g. ``num_qubits=2``.
+        """
+        super().set_params(**params)
+        if "num_qubits" in params:
+            self._left_qubits = [i for i in range(self.num_qubits)]
+            self._num_parameters = 0
```

Comment:
now, if we set both `num_qubits` and `num_params`, `num_params` will always be `0`. Is this wanted behavior?

Reply:
The number of parameters of the encoding circuit are determined automatically by the gates applied and therefore should not be set externally. The set_params function should not have a keyword "num_parameters". Therefore this should be appropriate, since when the number of qubits change it must be newly calculated how many parameters are needed to parametrize the given circuit.

Reply:
ah my bad, I mixed it up with `num_features`. You're totally right :)